### PR TITLE
Data interval calculation logic improvised

### PR DESCRIPTION
**Pull request by amarendra-dianomic**

### CODE REVIEW

File: python/foglamp/plugins/south/sinusoid/sinusoid.py [hunk 1]
```diff
@@ -166,35 +168,40 @@ def generate_data():
 
     async def save_data():
         try:
+            time_time = time.time
+            start = time_time()
+            recs = int(handle['dataPointsPerSec']['value'])
+            period = round(1.0 / recs, len(str(recs))+1)
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/sinusoid/sinusoid.py]: why the ZeroDivisionError code is not applied here as in the deleted code.

File: python/foglamp/plugins/south/sinusoid/sinusoid.py [hunk 2]
```diff
@@ -166,35 +168,40 @@ def generate_data():
...
 
-    _task = asyncio.ensure_future(save_data())
+    def run_task(loop):
+        global _task
+        asyncio.set_event_loop(loop)
+        _task = asyncio.ensure_future(save_data(), loop=loop)
 
+    new_loop = asyncio.get_event_loop()
+    t = Thread(target=run_task, args=(new_loop,))
+    t.start()
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/sinusoid/sinusoid.py]: could you please share the results of shutdwn and reconfigure?

**Review by praveen-garg — CHANGES_REQUESTED** [review]

File: python/foglamp/plugins/south/sinusoid/sinusoid.py [hunk 3]
```diff
@@ -166,35 +168,40 @@ def generate_data():
 
     async def save_data():
         try:
+            time_time = time.time
+            start = time_time()
+            recs = int(handle['dataPointsPerSec']['value'])
```
Comment (ashish-jabble) [inline python/foglamp/plugins/south/sinusoid/sinusoid.py]: you have changed everything which we fixed before, See https://github.com/foglamp/foglamp-south-sinusoid/pull/15/files

**Review by ashish-jabble — CHANGES_REQUESTED** [review]

**Comment by amarendra-dianomic** [comment]
@praveen-garg @ashish-jabble changes done as advised. Below is the excerpt of the syslog for reconfigure and shutdown both:
```
Sep 25 13:20:40 asinha-Vostro-3458 FogLAMP[2086] INFO: sinusoid: foglamp.plugins.south.sinusoid.sinusoid: Old config for sinusoid plugin {'assetName': {'default': 'sinusoid', 'value': 'sinusoid', 'description': 'Name of Asset', 'type': 'string', 'order': '1'}, 'plugin': {'default': 'sinusoid', 'description': 'Sinusoid Plugin', 'readonly': 'true', 'type': 'string', 'value': 'sinusoid'}, 'dataPointsPerSec': {'default': '1', 'value': '1', 'description': 'Data points per second', 'type': 'integer', 'order': '2'}} #012 new config {'assetName': {'default': 'sinusoid', 'value': 'sinusoid', 'description': 'Name of Asset', 'type': 'string', 'order': '1'}, 'plugin': {'default': 'sinusoid', 'readonly': 'true', 'description': 'Sinusoid Plugin', 'type': 'string', 'value': 'sinusoid'}, 'dataPointsPerSec': {'default': '1', 'value': '5000', 'description': 'Data points per second', 'type': 'integer', 'order': '2'}}
Sep 25 13:20:40 asinha-Vostro-3458 FogLAMP[2086] INFO: sinusoid: foglamp.plugins.south.sinusoid.sinusoid: sinusoid plugin shut down.
Sep 25 13:20:40 asinha-Vostro-3458 FogLAMP[2086] INFO: sinusoid: foglamp.plugins.south.sinusoid.sinusoid: Restarting Sinusoid plugin due to change in configuration key [dataPointsPerSec]
Sep 25 13:20:41 asinha-Vostro-3458 FogLAMP[1942] INFO: scheduler: foglamp.services.core.scheduler.scheduler: Process started: Schedule 'stats collection' process 'stats collector' task 3cde6898-83f6-4178-b622-83eb12a54f9d pid 3750, 2 running tasks#012['tasks/statistics', '--port=35497', '--address=127.0.0.1', '--name=stats collection']
Sep 25 13:20:41 asinha-Vostro-3458 FogLAMP[1942] INFO: scheduler: foglamp.services.core.scheduler.scheduler: Sleeping for 2658.141637802124 seconds
Sep 25 13:20:43 asinha-Vostro-3458 FogLAMP[1942] INFO: scheduler: foglamp.services.core.scheduler.scheduler: Process terminated: Schedule 'stats collection' process 'stats collector' task 3cde6898-83f6-4178-b622-83eb12a54f9d pid 3750 exit 0, 1 running tasks#012['tasks/statistics']
Sep 25 13:20:43 asinha-Vostro-3458 FogLAMP[1942] INFO: scheduler: foglamp.services.core.scheduler.scheduler: Scheduled task for schedule 'stats collection' to start at 2018-09-25 13:20:56.718185
Sep 25 13:20:43 asinha-Vostro-3458 FogLAMP[1942] INFO: scheduler: foglamp.services.core.scheduler.scheduler: Sleeping for 13.429284572601318 seconds
Sep 25 13:20:56 asinha-Vostro-3458 FogLAMP[1942] INFO: scheduler: foglamp.services.core.scheduler.scheduler: Process started: Schedule 'stats collection' process 'stats collector' task fd88d784-4b5e-49e2-9075-468250efb65f pid 28536, 2 running tasks#012['tasks/statistics', '--port=35497', '--address=127.0.0.1', '--name=stats collection']
Sep 25 13:20:56 asinha-Vostro-3458 FogLAMP[1942] INFO: scheduler: foglamp.services.core.scheduler.scheduler: Sleeping for 2643.1238067150116 seconds
Sep 25 13:20:58 asinha-Vostro-3458 FogLAMP[1942] INFO: scheduler: foglamp.services.core.scheduler.scheduler: Process terminated: Schedule 'stats collection' process 'stats collector' task fd88d784-4b5e-49e2-9075-468250efb65f pid 28536 exit 0, 1 running tasks#012['tasks/statistics']
Sep 25 13:20:58 asinha-Vostro-3458 FogLAMP[1942] INFO: scheduler: foglamp.services.core.scheduler.scheduler: Scheduled task for schedule 'stats collection' to start at 2018-09-25 13:21:11.718185
Sep 25 13:20:58 asinha-Vostro-3458 FogLAMP[1942] INFO: scheduler: foglamp.services.core.scheduler.scheduler: Sleeping for 13.618010997772217 seconds
Sep 25 13:21:09 asinha-Vostro-3458 FogLAMP[1942] INFO: common: foglamp.services.core.api.common: Executing controlled shutdown
Sep 25 13:21:09 asinha-Vostro-3458 FogLAMP[1942] INFO: scheduler: foglamp.services.core.scheduler.scheduler: Processing stop request
Sep 25 13:21:14 asinha-Vostro-3458 FogLAMP[1942] INFO: scheduler: foglamp.services.core.scheduler.scheduler: Stopped
Sep 25 13:21:14 asinha-Vostro-3458 FogLAMP[1942] INFO: server: foglamp.services.core.server: Shutting down the Southbound service s ...
Sep 25 13:21:14 asinha-Vostro-3458 FogLAMP[2086] INFO: sinusoid: foglamp.plugins.south.sinusoid.sinusoid: sinusoid plugin shut down.
Sep 25 13:21:22 asinha-Vostro-3458 FogLAMP[1942] INFO: service_registry: foglamp.services.core.service_registry.service_registry: Stopped service instance id=3f51cf1d-9d1e-4382-933d-7dd9a341da50: <s, type=Southbound, protocol=http, address=0.0.0.0, service port=45991, management port=45991, status=2>
Sep 25 13:21:24 asinha-Vostro-3458 FogLAMP[1942] INFO: server: foglamp.services.core.server: Successfully shut down the Southbound service s.
Sep 25 13:21:24 asinha-Vostro-3458 FogLAMP[1942] INFO: server: foglamp.services.core.server: Services monitoring stopped.
Sep 25 13:21:24 asinha-Vostro-3458 FogLAMP[1942] INFO: server: foglamp.services.core.server: Rest server stopped.
Sep 25 13:21:24 asinha-Vostro-3458 FogLAMP[1942] INFO: server: foglamp.services.core.server: Shutting down the Storage service FogLAMP Storage ...
Sep 25 13:21:24 asinha-Vostro-3458 FogLAMP Storage[1947]: Storage service shutdown in progress.
Sep 25 13:21:24 asinha-Vostro-3458 FogLAMP[1942] INFO: server: foglamp.services.core.server: Successfully shut down the Storage service FogLAMP Storage.
Sep 25 13:21:24 asinha-Vostro-3458 FogLAMP[1942] INFO: server: foglamp.services.core.server: FogLAMP PID file [/home/asinha/Development/FogLAMP/data/var/run/foglamp.core.pid] removed.
Sep 25 13:21:24 asinha-Vostro-3458 FogLAMP[1942] INFO: service_registry: foglamp.services.core.service_registry.service_registry: Stopped service instance id=7d4e5263-2b10-42fa-a24b-1a524692b46f: <FogLAMP Storage, type=Storage, protocol=http, address=localhost, service port=43169, management port=33433, status=2>
Sep 25 13:21:24 asinha-Vostro-3458 FogLAMP Storage[1947]: Unregistered service 7d4e5263-2b10-42fa-a24b-1a524692b46f.
Sep 25 13:21:24 asinha-Vostro-3458 FogLAMP Storage[1947]: Storage service shut down.
Sep 25 13:21:26 asinha-Vostro-3458 FogLAMP[1942] INFO: server: foglamp.services.core.server: Stopping the FogLAMP Core event loop. Good Bye!

```

File: python/foglamp/plugins/south/sinusoid/sinusoid.py (hunk 3 above)
Comment (amarendra-dianomic) [reply python/foglamp/plugins/south/sinusoid/sinusoid.py]: restored

File: python/foglamp/plugins/south/sinusoid/sinusoid.py (hunk 2 above)
Comment (amarendra-dianomic) [reply python/foglamp/plugins/south/sinusoid/sinusoid.py]: done

File: python/foglamp/plugins/south/sinusoid/sinusoid.py (hunk 1 above)
Comment (amarendra-dianomic) [reply python/foglamp/plugins/south/sinusoid/sinusoid.py]: restored

**Review by praveen-garg — DISMISSED** [review]
LGTM!

**Review by Singhal-Vaibhav — CHANGES_REQUESTED** [review]
Configured sinusoid plugin, and changed dataPointsPerSec, observation as follows:
1. value to 1000 -> Plugin stable and able to ingest
2. changed from 1000 to 10000 -> Plugin unstable ```WARNING: ingest: foglamp.services.south.ingest: The ingest service is unavailable ```
3. changed from 10000 to 1000 -> Plugin still unstable and same message is repeated.

**Comment by amarendra-dianomic** [comment]
@Singhal-Vaibhav errors fixed. To test with "develop" branch of FogLAMP, ensure that Ingest queue size is set to max dps rate you intend to test with. Otherwise, a companion [PR #1144](https://github.com/foglamp/FogLAMP/pull/1144) has been created to auto adjust the queue size and also to incorporate a few other refactorings.

**Comment by amarendra-dianomic** [comment]
@Singhal-Vaibhav, @praveen-garg  code fixed for reconfigure problem due to asset name change or data points per sec change. Now previous asset is not incrementing on asset name change. Tested with both asset name change and data rate change simultaneously for changing data rate from 1000 to 10000 to 1000 to 1, along with asset name change and everything worked smoothly. Pl check.

File: python/foglamp/plugins/south/sinusoid/sinusoid.py [hunk 4]
```diff
@@ -166,35 +168,48 @@ def generate_data():
...
             pass
         except (Exception, RuntimeError) as ex:
             _LOGGER.exception("Sinusoid exception: {}".format(str(ex)))
             raise exceptions.DataRetrievalError(ex)
 
-    _task = asyncio.ensure_future(save_data())
-
+    def run_task(loop):
+        global _task, should_stop
+        # _LOGGER.warning('>>>>>>> No. of active threads: %d, time: %s', active_count(), time.time())
```
Comment (praveen-garg) [inline python/foglamp/plugins/south/sinusoid/sinusoid.py]: Remove it, or comment with `_LOGGER.info('No. of active threads: %d, time: %s', active_count(), time.time())`

**Comment by praveen-garg** [comment]
let's close this and re-open with fixes if required.

**Pull request closed without merging**